rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+
+This RFC proposes to introduce centralized configurations on Meta Service and data will be persisted in etcd.
+
+For the sake of clarity, in this doc, we will use the term **"parameters"** for system and session parameters, and **"config"** for command-line arguments or config files.
+
+**Note:** Previously I think this might not be an urgent issue until I realized that our Cloud platform couples tightly with more and more configurations. The later we commence, the heavier burden it will become. 
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Parameters (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Parameters
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
+
+Here we could borrow the syntax from Postgres:
+
+```sql
+-- Set or reset a parameter
+ALTER SYSTEM / DATABASE SET parameter TO value / 'value' / DEFAULT;
+
+-- List all parameters keys & values
+SHOW PARAMETERS;
+```
+
+> **Note**: PG uses `SHOW ALL` because it doesn't has other `SHOW` commands like us. Which one is better?
+
+
+References:
+
+- [PostgreSQL: Documentation: 15: ALTER SYSTEM](https://www.postgresql.org/docs/current/sql-altersystem.html)
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+- [PostgreSQL: Documentation: 15: SHOW](https://www.postgresql.org/docs/current/sql-show.html)
+
+Similar to other system metadata e.g. catalog,
+
+- Changes to global parameters should be propagated to all nodes, in the cluster in a deterministic order ideally
+- When a node reboots or recovers, the latest snapshot of parameters should be passed from Meta Service.
+
+Therefore, the Notification Service seems to fit the scene well.
+
+### Session Parameters
+
+PostgreSQL allows users to set parameters for the current session as well.
+
+```sql
+SET parameter TO value / 'value' / DEFAULT;
+```
+
+References:
+
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+
+Previously, we had already introduced [session parameters in RisingWave](https://github.com/risingwavelabs/risingwave/blob/53f7e0db772ac7e51773791bb8301624ed763ae8/src/common/src/session_config/mod.rs#L265). Some of them should be system parameters like `batch_enable_sort_agg` or `query_mode`. Postgres **enforces** that session parameters must also be system paramters, we can follow this rule as well.
```

Comment:
Session parameters seems to conflict with the definition of system parameters, since the values can vary from session to session (on different fe nodes). So I think session parameters should be out of scope for this RFC.